hs_core/hydroshare/resource.py

```diff
@@ -106,8 +136,61 @@ def update_quota_usage(username):
         logger.error(err_msg)
         raise ValidationError(err_msg)
 
-    used_val = get_quota_usage_from_irods(username)
-    uq.update_used_value(used_val)
+    uz, dz = get_quota_usage_from_irods(username, raise_on_error=False)
+
+    # subtract out published resources from the datazone
+    original_quota_data = get_quota_data(uq)
+    qmsg = original_quota_data["qmsg"]
+    published_percent = qmsg.published_resource_percent
+    user = User.objects.get(username=username)
+    published_size = get_storage_usage(user, flag="published")
+    dz -= published_size * (1 - published_percent)
+    uq.update_used_value(uz, dz)
+
+    if original_quota_data["enforce_quota"]:
+        updated_quota_data = get_quota_data(uq)
+        # if enforcing quota, take steps to send messages
+        percent = updated_quota_data["percent"]
+        if percent < qmsg.soft_limit_percent:
+            # No need for further action
+            return
+        today = datetime.date.today()
+        if percent >= qmsg.soft_limit_percent:
+            if percent >= 100 and percent < qmsg.hard_limit_percent:
+                if not uq.grace_period_ends:
+                    # triggers grace period counting
+                    uq.grace_period_ends = today + datetime.timedelta(days=qmsg.grace_period)
+                    tasks.send_user_notification_at_quota_grace_start.apply_async((user.pk))
+            elif percent >= qmsg.hard_limit_percent:
+                # reset grace period when user quota exceeds hard limit
+                uq.grace_period_ends = None
+            uq.save()
+        else:
+            if uq.grace_period_ends and uq.grace_period_ends < today :
+                # reset grace period now that the user is below quota soft limit
+                uq.grace_period_ends = None
+                uq.save()
+        updated_quota_status = updated_quota_data["status"]
+        original_quota_status = original_quota_data["status"]
+        if original_quota_status == QuotaStatus.ENFORCEMENT and updated_quota_status == QuotaStatus.ENFORCEMENT:
```

Comment:
this is because we only want to notify admin if a user is already in quota enforcement -- the infraction is that they have already been notified that they are over quota and they are continuing to upload to the userZone

Reply:
I refactored this into a separate function to maybe make it more clear